TheJumpCloud/support
# Sa 2539 update jc module fix

## Issues
* [SA-2539](https://jumpcloud.atlassian.net/browse/SA-2539) - Update-JCModule is not updating SDKs as expected

## What does this solve?

When running Update-JCModule customers would be presented with an ugly error message which states that 'x' sdk module could not be uninstalled.

## Is there anything particularly tricky?

No, this should be easy to test as long as the guidelines below are followed.

## How should this be tested?

**Test the issue in production by slightly modifying Update-JCModule**

First ensure that you have the latest public version of the SDKs installed:
`
Install-Module -Name JumpCloud.SDK.DirectoryInsights -RequiredVersion 0.0.15 -Force
Install-Module -Name JumpCloud.SDK.V2 -RequiredVersion 0.0.31 -Force
Install-Module -Name JumpCloud.SDK.V1 -RequiredVersion 0.0.27 -Force
`
Uninstall later versions if they exist.
Go to your local PowerShell Folder Location: `~/.local/share/powershell/Modules`
Navigate to Jumpcloud/1.20.1/Public/Authentication
Open up Update-JCModule.ps1 and edit lines 23-30 which look like:
`Else
  {
      $FoundModule = Find-Module -Name:($ModuleName) -Repository:($Repository)
      $FoundSDKs = foreach ($SDK in $SDKs)
      {
          Find-Module -Name:($SDK) -Repository:($Repository)
      }
  }
`
**replace** with specific versions of the sdks (these are hidden until we fix this but you can still locate them on PSGallery if you know the version):
`
Else
{
    $FoundModule = Find-Module -Name:($ModuleName) -Repository:($Repository)
    $FoundSDKs = foreach ($SDK in $SDKs)
    {
        if ($SDK -eq 'JumpCloud.SDK.DirectoryInsights'){
            Find-Module -Name:($SDK) -Repository:($Repository) -RequiredVersion '0.0.16'
        }
        if ($SDK -eq 'JumpCloud.SDK.V2'){
            Find-Module -Name:($SDK) -Repository:($Repository) -RequiredVersion '0.0.32'
        }
        if ($SDK -eq 'JumpCloud.SDK.V1'){
            Find-Module -Name:($SDK) -Repository:($Repository) -RequiredVersion '0.0.28'
        }
    }
}
`
Run `Update-JCModule -Debug` and validate that the function fails to update the sdks.

**Test the change in this PR**

First ensure that you have the latest public version of the SDKs installed:
`
Install-Module -Name JumpCloud.SDK.DirectoryInsights -RequiredVersion 0.0.15 -Force
Install-Module -Name JumpCloud.SDK.V2 -RequiredVersion 0.0.31 -Force
Install-Module -Name JumpCloud.SDK.V1 -RequiredVersion 0.0.27 -Force
`
Uninstall later versions if they exist.
Go to your local PowerShell Folder Location: `~/.local/share/powershell/Modules`
Navigate to Jumpcloud/1.20.1/Public/Authentication
Open up Update-JCModule.ps1 replace contents with [latest version from this branch](https://github.com/TheJumpCloud/support/blob/SA-2539_updateJCModuleFix/PowerShell/JumpCloud%20Module/Public/Authentication/Update-JCModule.ps1)

Run `Update-JCModule -Debug` and validate that the function updates the sdks and prompts user to restart their powershell session.


## Screenshots

Current Error
![Screen Shot 2022-05-03 at 2 23 53 PM](https://user-images.githubusercontent.com/54448601/166570808-ef809fdf-4e8b-43c7-a13f-284a49a4d6d6.png)

Fixed Workflow
![Screen Shot 2022-05-03 at 3 38 04 PM](https://user-images.githubusercontent.com/54448601/166570727-95ad5c99-9ad9-4071-a4d7-6475fae76cbc.png)
